File: zig/persistent_merkle_tree.zig
const std = @import("std");
const pmt = @import("ssz:persistent_merkle_tree");
const toErrCode = @import("common.zig").toErrCode;

var allocator = std.heap.GeneralPurposeAllocator(.{}){};
var pool: pmt.Node.Pool = undefined;

// Pool

pub export fn persistent_merkle_tree_pool_init(pool_size: u32) i32 {
    pool = pmt.Node.Pool.init(allocator.allocator(), pool_size) catch |e| return toErrCode(e);
    return 0;
}

pub export fn persistent_merkle_tree_pool_deinit() void {
    pool.deinit();
}

pub export fn persistent_merkle_tree_pool_create_leaf(hash: [*c]u8, should_ref: bool) i32 {
    const id = pmt.Node.Pool.createLeaf(
        &pool,
        hash[0..32],
        should_ref,
    ) catch |e| return toErrCode(e);
    return @intCast(@intFromEnum(id));
}

pub export fn persistent_merkle_tree_pool_create_branch(left_id: u32, right_id: u32, should_ref: bool) i32 {
    const id = pmt.Node.Pool.createBranch(
        &pool,
        @enumFromInt(left_id),
        @enumFromInt(right_id),
        should_ref,
    ) catch |e| return toErrCode(e);
    return @intCast(@intFromEnum(id));
}

pub export fn persistent_merkle_tree_pool_ref(id: u32) i32 {
    _ = pmt.Node.Pool.ref(&pool, @enumFromInt(id)) catch |e| return toErrCode(e);
    return 0;
}

pub export fn persistent_merkle_tree_pool_unref(id: u32) void {
    pmt.Node.Pool.unref(&pool, @enumFromInt(id));
}

// Node - basic

pub export fn persistent_merkle_tree_node_get_left(id: u32) i32 {
    const node = pmt.Node.Id.getLeft(@enumFromInt(id), &pool) catch |e| return toErrCode(e);
    return @intCast(@intFromEnum(node));
}

pub export fn persistent_merkle_tree_node_get_right(id: u32) i32 {
    const node = pmt.Node.Id.getRight(@enumFromInt(id), &pool) catch |e| return toErrCode(e);
    return @intCast(@intFromEnum(node));
}

pub export fn persistent_merkle_tree_node_get_hash(id: u32) [*c]const u8 {
    return @ptrCast(pmt.Node.Id.getRoot(@enumFromInt(id), &pool));
}

pub export fn persistent_merkle_tree_node_get_state(id: u32) u32 {
    const state = pmt.Node.Id.getState(@enumFromInt(id), &pool);
    return @intFromEnum(state);
}

// Node - navigation

pub export fn persistent_merkle_tree_node_get_node(root_id: u32, gindex: u64) i32 {
    const id = pmt.Node.Id.getNode(
        @enumFromInt(root_id),
        &pool,
        @enumFromInt(gindex),
    ) catch |e| return toErrCode(e);
    return @intCast(@intFromEnum(id));
}

pub export fn persistent_merkle_tree_node_get_node_at_depth(root_id: u32, depth: u8, index: u32) i32 {
    const id = pmt.Node.Id.getNodeAtDepth(
        @enumFromInt(root_id),
        &pool,
        @intCast(depth),
        index,
    ) catch |e| return toErrCode(e);
    return @intCast(@intFromEnum(id));
}

pub export fn persistent_merkle_tree_node_get_nodes_at_depth(root_id: u32, depth: u8, start_index: u32, out: [*c]u32, out_len: u32) i32 {
    pmt.Node.Id.getNodesAtDepth(
        @enumFromInt(root_id),
        &pool,
        @intCast(depth),
        start_index,
        @ptrCast(out[0..out_len]),
    ) catch |e| return toErrCode(e);
    return 0;
}

pub export fn persistent_merkle_tree_node_set_node(root_id: u32, gindex: u64, node_id: u32) i32 {
    const new_root_id = pmt.Node.Id.setNode(
        @enumFromInt(root_id),
        &pool,
        @enumFromInt(gindex),
        @enumFromInt(node_id),
    ) catch |e| return toErrCode(e);
    return @intCast(@intFromEnum(new_root_id));
}

pub export fn persistent_merkle_tree_node_set_node_at_depth(root_id: u32, depth: u8, index: u32, node_id: u32) i32 {
    const new_root_id = pmt.Node.Id.setNodeAtDepth(
        @enumFromInt(root_id),
        &pool,
        @intCast(depth),
        index,
        @enumFromInt(node_id),
    ) catch |e| return toErrCode(e);
    return @intCast(@intFromEnum(new_root_id));
}

pub export fn persistent_merkle_tree_node_set_nodes_at_depth(root_id: u32, depth: u8, indices: [*c]const u64, node_ids: [*c]u32, len: u32) i32 {
    const new_root_id = pmt.Node.Id.setNodesAtDepth(
        @enumFromInt(root_id),
        &pool,
        @intCast(depth),
        @ptrCast(indices[0..len]),
        @ptrCast(node_ids[0..len]),
    ) catch |e| return toErrCode(e);
    return @intCast(@intFromEnum(new_root_id));
}

pub export fn persistent_merkle_tree_node_set_nodes(root_id: u32, gindices: [*c]const u64, node_ids: [*c]u32, len: u32) i32 {
    const new_root_id = pmt.Node.Id.setNodes(
        @enumFromInt(root_id),
        &pool,
        @ptrCast(gindices[0..len]),
        @ptrCast(node_ids[0..len]),
    ) catch |e| return toErrCode(e);
    return @intCast(@intFromEnum(new_root_id));
}

// Node - tree creation

pub export fn persistent_merkle_tree_node_fill_to_depth(leaf: u32, depth: u8, should_ref: bool) i32 {
    const root_id = pmt.Node.fillToDepth(
        &pool,
        @enumFromInt(leaf),
        @intCast(depth),
        should_ref,
    ) catch |e| return toErrCode(e);
    return @intCast(@intFromEnum(root_id));
}

pub export fn persistent_merkle_tree_node_fill_to_length(leaf: u32, depth: u8, length: u32, should_ref: bool) i32 {
    const root_id = pmt.Node.fillToLength(
        &pool,
        @enumFromInt(leaf),
        @intCast(depth),
        length,
        should_ref,
    ) catch |e| return toErrCode(e);
    return @intCast(@intFromEnum(root_id));
}

pub export fn persistent_merkle_tree_node_fill_with_contents(leaf_ids: [*c]u32, len: u32, depth: u8, should_ref: bool) i32 {
    const root_id = pmt.Node.fillWithContents(
        &pool,
        @ptrCast(leaf_ids[0..len]),
        @intCast(depth),
        should_ref,
    ) catch |e| return toErrCode(e);
    return @intCast(@intFromEnum(root_id));
}

File: src/persistent_merkle_tree.ts
import { toArrayBuffer, type Pointer } from "bun:ffi";
import {binding} from "./binding.ts";
import { throwErr } from "./common.ts";

export function poolInit(size: number): void {
  throwErr(binding.persistent_merkle_tree_pool_init(size));
}

export function poolDeinit(): void {
  binding.persistent_merkle_tree_pool_deinit();
}

export type Node = number & {["persistent_merkle_tree_node"]: never};

export function createLeaf(data: Uint8Array, shouldRef = false): Node {
  if (data.length !== 32) {
    throw new Error("Leaf data must be 32 bytes");
  }
  return throwErr(binding.persistent_merkle_tree_pool_create_leaf(data, shouldRef)) as Node;
}

export function createBranch(left: Node, right: Node, shouldRef = false): Node {
  return throwErr(binding.persistent_merkle_tree_pool_create_branch(left, right, shouldRef)) as Node;
}

export function ref(node: Node): void {
  throwErr(binding.persistent_merkle_tree_pool_ref(node));
}

export function unref(node: Node): void {
  binding.persistent_merkle_tree_pool_unref(node);
}

export function getLeft(node: Node): Node {
  return throwErr(binding.persistent_merkle_tree_node_get_left(node)) as Node;
}

export function getRight(node: Node): Node {
  return throwErr(binding.persistent_merkle_tree_node_get_right(node)) as Node;
}

export function getHash(node: Node): Uint8Array {
  return new Uint8Array(toArrayBuffer(binding.persistent_merkle_tree_node_get_hash(node) as Pointer, 0, 32));
}

export enum NodeType {
  Zero = 0x00000000,
  Leaf = 0x20000000,
  BranchLazy = 0x40000000,
  BranchComputed = 0x60000000,
}

export type NodeState = {
  type: NodeType;
  refCount: number;
}

export function getState(node: Node): NodeState {
  const state = binding.persistent_merkle_tree_node_get_state(node);
  return {
    type: state & 0x60000000,
    refCount: state & 0x1fffffff,
  };
}

export function getNode(rootNode: Node, gindex: number | bigint): Node {
  return throwErr(binding.persistent_merkle_tree_node_get_node(rootNode, gindex)) as Node;
}

export function getNodeAtDepth(rootNode: Node, depth: number, index: number): Node {
  return throwErr(binding.persistent_merkle_tree_node_get_node_at_depth(rootNode, depth, index)) as Node;
}

export function getNodesAtDepth(rootNode: Node, depth: number, startIndex: number, count: number): Uint32Array {
  const nodes = new Uint32Array(count);
  throwErr(binding.persistent_merkle_tree_node_get_nodes_at_depth(rootNode, depth, startIndex, nodes, count));
  return nodes;
}

export function setNode(rootNode: Node, gindex: number | bigint, newNode: Node): Node {
  return throwErr(binding.persistent_merkle_tree_node_set_node(rootNode, gindex, newNode)) as Node;
}

export function setNodeAtDepth(rootNode: Node, depth: number, index: number, newNode: Node): Node {
  return throwErr(binding.persistent_merkle_tree_node_set_node_at_depth(rootNode, depth, index, newNode)) as Node;
}

export function setNodesAtDepth(rootNode: Node, depth: number, indices: number[], newNodes: Node[]): Node {
  if (indices.length !== newNodes.length) {
    throw new Error("Indices and newNodes must have the same length");
  }
  return throwErr(binding.persistent_merkle_tree_node_set_nodes_at_depth(
    rootNode,
    depth,
    new BigUint64Array(indices.map(BigInt)),
    new Uint32Array(newNodes),
    newNodes.length,
  )) as Node;
}

export function setNodes(rootNode: Node, gindices: (number | bigint)[], newNodes: Node[]): Node {
  if (gindices.length !== newNodes.length) {
    throw new Error("Gindices and newNodes must have the same length");
  }
  return throwErr(binding.persistent_merkle_tree_node_set_nodes(
    rootNode,
    new BigUint64Array(gindices.map(BigInt)),
    new Uint32Array(newNodes),
    newNodes.length,
  )) as Node;
}

export function fillToDepth(leaf: Node, depth: number, shouldRef = false): Node {
  return throwErr(binding.persistent_merkle_tree_node_fill_to_depth(leaf, depth, shouldRef)) as Node;
}

export function fillToLength(leaf: Node, depth: number, length: number, shouldRef = false): Node {
  return throwErr(binding.persistent_merkle_tree_node_fill_to_length(leaf, depth, length, shouldRef)) as Node;
}

export function fillWithContents(leaves: Node[], depth: number, shouldRef = false): Node {
  return throwErr(binding.persistent_merkle_tree_node_fill_with_contents(
    new Uint32Array(leaves),
    leaves.length,
    depth,
    shouldRef,
  )) as Node;
}